From: Boris Fiuczynski <fiuczy@linux.ibm.com>
Date: Wed, 6 Nov 2024 08:51:03 +0100
Subject: qemu: command: add multi boot device support on s390x

If QEMU supports multi boot device make use of it instead of using the
single boot device machine parameter.

Signed-off-by: Boris Fiuczynski <fiuczy@linux.ibm.com>
Reviewed-by: Peter Krempa <pkrempa@redhat.com>

Backported by dropping the new tests around Qemu 9.1.0, which is not available
in Ubuntu Noble/Oracular.

Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+bug/2051239
Origin: backport, https://gitlab.com/libvirt/libvirt/-/commit/bf0308b2d4
---
 src/qemu/qemu_command.c | 36 ++++++++++++++++++++++++++++++------
 src/qemu/qemu_command.h |  6 ++++--
 src/qemu/qemu_hotplug.c |  6 ++++--
 3 files changed, 38 insertions(+), 10 deletions(-)

diff --git a/src/qemu/qemu_command.c b/src/qemu/qemu_command.c
index 6538171..6ace65a 100644
--- a/src/qemu/qemu_command.c
+++ b/src/qemu/qemu_command.c
@@ -1782,6 +1782,7 @@ qemuBuildDiskDeviceProps(const virDomainDef *def,
     g_autofree char *chardev = NULL;
     g_autofree char *drive = NULL;
     unsigned int bootindex = 0;
+    const char *bootLoadparm = NULL;
     unsigned int logical_block_size = disk->blockio.logical_block_size;
     unsigned int physical_block_size = disk->blockio.physical_block_size;
     unsigned int discard_granularity = disk->blockio.discard_granularity;
@@ -1920,9 +1921,13 @@ qemuBuildDiskDeviceProps(const virDomainDef *def,
     }
 
     /* bootindex for floppies is configured via the fdc controller */
-    if (disk->device != VIR_DOMAIN_DISK_DEVICE_FLOPPY)
+    if (disk->device != VIR_DOMAIN_DISK_DEVICE_FLOPPY) {
         bootindex = disk->info.effectiveBootIndex;
 
+        if (virQEMUCapsGet(qemuCaps, QEMU_CAPS_VIRTIO_CCW_DEVICE_LOADPARM))
+            bootLoadparm = disk->info.loadparm;
+    }
+
     if (disk->wwn) {
         unsigned long long w = 0;
 
@@ -1965,6 +1970,7 @@ qemuBuildDiskDeviceProps(const virDomainDef *def,
                               "S:chardev", chardev,
                               "s:id", disk->info.alias,
                               "p:bootindex", bootindex,
+                              "S:loadparm", bootLoadparm,
                               "p:logical_block_size", logical_block_size,
                               "p:physical_block_size", physical_block_size,
                               "p:discard_granularity", discard_granularity,
@@ -3821,6 +3827,7 @@ qemuBuildNicDevProps(virDomainDef *def,
     g_autoptr(virJSONValue) props = NULL;
     char macaddr[VIR_MAC_STRING_BUFLEN];
     g_autofree char *netdev = g_strdup_printf("host%s", net->info.alias);
+    const char *bootLoadparm = NULL;
 
     if (virDomainNetIsVirtioModel(net)) {
         const char *tx = NULL;
@@ -3906,11 +3913,15 @@ qemuBuildNicDevProps(virDomainDef *def,
 
     virMacAddrFormat(&net->mac, macaddr);
 
+    if (virQEMUCapsGet(qemuCaps, QEMU_CAPS_VIRTIO_CCW_DEVICE_LOADPARM))
+        bootLoadparm = net->info.loadparm;
+
     if (virJSONValueObjectAdd(&props,
                               "s:netdev", netdev,
                               "s:id", net->info.alias,
                               "s:mac", macaddr,
                               "p:bootindex", net->info.effectiveBootIndex,
+                              "S:loadparm", bootLoadparm,
                               NULL) < 0)
         return NULL;
 
@@ -4930,15 +4941,21 @@ qemuBuildSCSIVHostHostdevDevProps(const virDomainDef *def,
 virJSONValue *
 qemuBuildSCSIHostdevDevProps(const virDomainDef *def,
                              virDomainHostdevDef *dev,
-                             const char *backendAlias)
+                             const char *backendAlias,
+                             virQEMUCaps *qemuCaps)
 {
+    const char *bootLoadparm = NULL;
     g_autoptr(virJSONValue) props = NULL;
 
+    if (virQEMUCapsGet(qemuCaps, QEMU_CAPS_VIRTIO_CCW_DEVICE_LOADPARM))
+        bootLoadparm = dev->info->loadparm;
+
     if (virJSONValueObjectAdd(&props,
                               "s:driver", "scsi-generic",
                               "s:drive", backendAlias,
                               "s:id", dev->info->alias,
                               "p:bootindex", dev->info->bootIndex,
+                              "S:loadparm", bootLoadparm,
                               NULL) < 0)
         return NULL;
 
@@ -5019,16 +5036,21 @@ qemuBuildHostdevMdevModelTypeString(virDomainHostdevSubsysMediatedDev *mdev)
 
 virJSONValue *
 qemuBuildHostdevMediatedDevProps(const virDomainDef *def,
-                                 virDomainHostdevDef *dev)
+                                 virDomainHostdevDef *dev,
+                                 virQEMUCaps *qemuCaps)
 {
     g_autoptr(virJSONValue) props = NULL;
     virDomainHostdevSubsysMediatedDev *mdevsrc = &dev->source.subsys.u.mdev;
     g_autofree char *mdevPath = NULL;
+    const char *bootLoadparm = NULL;
     /* 'ramfb' property must be omitted unless it's to be enabled */
     bool ramfb = mdevsrc->ramfb == VIR_TRISTATE_SWITCH_ON;
 
     mdevPath = virMediatedDeviceGetSysfsPath(mdevsrc->uuidstr);
 
+    if (virQEMUCapsGet(qemuCaps, QEMU_CAPS_VIRTIO_CCW_DEVICE_LOADPARM))
+        bootLoadparm = dev->info->loadparm;
+
     if (virJSONValueObjectAdd(&props,
                               "s:driver", qemuBuildHostdevMdevModelTypeString(mdevsrc),
                               "s:id", dev->info->alias,
@@ -5036,6 +5058,7 @@ qemuBuildHostdevMediatedDevProps(const virDomainDef *def,
                               "S:display", qemuOnOffAuto(mdevsrc->display),
                               "B:ramfb", ramfb,
                               "p:bootindex", dev->info->bootIndex,
+                              "S:loadparm", bootLoadparm,
                               NULL) < 0)
         return NULL;
 
@@ -5134,7 +5157,7 @@ qemuBuildHostdevSCSICommandLine(virCommand *cmd,
     if (qemuBuildBlockStorageSourceAttachDataCommandline(cmd, data, qemuCaps) < 0)
         return -1;
 
-    if (!(devprops = qemuBuildSCSIHostdevDevProps(def, hostdev, backendAlias)))
+    if (!(devprops = qemuBuildSCSIHostdevDevProps(def, hostdev, backendAlias, qemuCaps)))
         return -1;
 
     if (qemuBuildDeviceCommandlineFromJSON(cmd, devprops, def, qemuCaps) < 0)
@@ -5233,7 +5256,7 @@ qemuBuildHostdevCommandLine(virCommand *cmd,
                 return -1;
             }
 
-            if (!(devprops = qemuBuildHostdevMediatedDevProps(def, hostdev)))
+            if (!(devprops = qemuBuildHostdevMediatedDevProps(def, hostdev, qemuCaps)))
                 return -1;
 
             if (qemuBuildDeviceCommandlineFromJSON(cmd, devprops, def, qemuCaps) < 0)
@@ -7030,7 +7053,8 @@ qemuBuildMachineCommandLine(virCommand *cmd,
         virBufferAsprintf(&buf, ",max-cpu-compat=%s", cpu->model);
     }
 
-    qemuAppendLoadparmMachineParm(&buf, def);
+    if (!virQEMUCapsGet(qemuCaps, QEMU_CAPS_VIRTIO_CCW_DEVICE_LOADPARM))
+        qemuAppendLoadparmMachineParm(&buf, def);
 
     if (def->sec) {
         switch ((virDomainLaunchSecurity) def->sec->sectype) {
diff --git a/src/qemu/qemu_command.h b/src/qemu/qemu_command.h
index 341ec43..1989389 100644
--- a/src/qemu/qemu_command.h
+++ b/src/qemu/qemu_command.h
@@ -178,7 +178,8 @@ qemuBuildUSBHostdevDevProps(const virDomainDef *def,
 virJSONValue *
 qemuBuildSCSIHostdevDevProps(const virDomainDef *def,
                              virDomainHostdevDef *dev,
-                             const char *backendAlias);
+                             const char *backendAlias,
+                             virQEMUCaps *qemuCaps);
 
 qemuBlockStorageSourceAttachData *
 qemuBuildHostdevSCSIAttachPrepare(virDomainHostdevDef *hostdev,
@@ -196,7 +197,8 @@ qemuBuildSCSIVHostHostdevDevProps(const virDomainDef *def,
 
 virJSONValue *
 qemuBuildHostdevMediatedDevProps(const virDomainDef *def,
-                                 virDomainHostdevDef *dev);
+                                 virDomainHostdevDef *dev,
+                                 virQEMUCaps *qemuCaps);
 
 virJSONValue *
 qemuBuildRedirdevDevProps(const virDomainDef *def,
diff --git a/src/qemu/qemu_hotplug.c b/src/qemu/qemu_hotplug.c
index 0e45bd5..4a7dc96 100644
--- a/src/qemu/qemu_hotplug.c
+++ b/src/qemu/qemu_hotplug.c
@@ -2500,7 +2500,8 @@ qemuDomainAttachHostSCSIDevice(virQEMUDriver *driver,
                                                    priv->qemuCaps)))
         goto cleanup;
 
-    if (!(devprops = qemuBuildSCSIHostdevDevProps(vm->def, hostdev, backendalias)))
+    if (!(devprops = qemuBuildSCSIHostdevDevProps(vm->def, hostdev, backendalias,
+                                                  priv->qemuCaps)))
         goto cleanup;
 
     VIR_REALLOC_N(vm->def->hostdevs, vm->def->nhostdevs + 1);
@@ -2720,7 +2721,8 @@ qemuDomainAttachMediatedDevice(virQEMUDriver *driver,
 
     qemuAssignDeviceHostdevAlias(vm->def, &hostdev->info->alias, -1);
 
-    if (!(devprops = qemuBuildHostdevMediatedDevProps(vm->def, hostdev)))
+    if (!(devprops = qemuBuildHostdevMediatedDevProps(vm->def, hostdev,
+                                                      priv->qemuCaps)))
         goto cleanup;
 
     VIR_REALLOC_N(vm->def->hostdevs, vm->def->nhostdevs + 1);
